compiler/rustc_serialize/src/opaque.rs

```diff
@@ -380,6 +381,38 @@ impl FileEncoder {
 
         Ok(())
     }
+
+    #[inline]
+    pub fn write_mmap<T: MmapSafe>(&mut self, val: &T) -> Result<usize, io::Error> {
+        self.write_mmap_slice(std::slice::from_ref(val))
+    }
+
+    #[inline]
+    pub fn write_mmap_slice<T: MmapSafe>(&mut self, val: &[T]) -> Result<usize, io::Error> {
+        const ALIGN: [u8; 64] = [0; 64];
+        // First, we need to align the memory.
+        let pos = self.position();
+        let align = std::mem::align_of::<T>();
+        let mask = align - 1;
+        let extra = pos & mask;
+        let padding = if extra == 0 { 0 } else { align - extra };
+        self.write_all(&ALIGN[..padding])?;
```

Comment:
Maybe skip this call altogether when padding is 0?

Reply:
I'm also wondering if it would be better to just assert that the data will be aligned, instead of adding padding. Then would just add some padding before the entire array? But I'm not sure if that's actually better.

Reply:
In general, we can't assert alignment. When encoding a `Fingerprint` (align 8) after `SerializedDepNodeIndex`es (align 4), there is no reason we should expect the encoder to be aligned.

Reply:
Yeah, that's a good point.